src/ui/PromotedResultsBadge/PromotedResultsBadge.ts

```diff
@@ -0,0 +1,187 @@
+  static ID = 'PromotedResultsBadge';
+
+  static doExport = () => {
+    exportGlobally({
+      PromotedResultsBadge
+    });
+  };
+
+  /**
+   * @componentOptions
+   */
+  static options: IPromotedResultsBadgeOptions = {
+    /**
+     * Specifies if a badge should be added to "Featured Results" configured through a [Coveo Query Pipeline](http://www.coveo.com/go?dest=cloudhelp&lcid=9&context=126).
+     *
+     * Default value is `true`.
+     */
+    showBadgeForFeaturedResults: ComponentOptions.buildBooleanOption({ defaultValue: true }),
+    /**
+     * Specifies if a badge should be added to "Recommended Results" returned by a [Coveo Machine Learning algorithm](http://www.coveo.com/go?dest=cloudhelp&lcid=9&context=183).
+     *
+     * Default value is `false`.
+     */
+    showBadgeForRecommendedResults: ComponentOptions.buildBooleanOption({ defaultValue: false }),
+
+    /**
+     * Specifies the caption that should be used for "Recommended Results".
+     *
+     * Default value is the localized string for `Recommended`.
+     */
+    captionForRecommended: ComponentOptions.buildLocalizedStringOption({
+      defaultValue: l('Recommended'),
+      depend: 'showBadgeForRecommendedResults'
+    }),
+    /**
+     * Specifies the caption that should be used for "Featured Results".
+     *
+     * Default value is the localized string for `Featured`.
+     */
+    captionForFeatured: ComponentOptions.buildLocalizedStringOption({ defaultValue: l('Featured'), depend: 'showBadgeForFeaturedResults' }),
+
+    /**
+     * Specifies the color that should be used for "Featured Results".
+     *
+     * This can be specified using either an hexadecimal value (eg: `#f58020`), an RGB value (eg: `rgb(125,10,36)`), or a CSS color name (eg: `pink`, `yellow`, `orange`, etc.).
+     *
+     * Default value is `undefined`, and is controlled through the default stylesheet of the framework.
+     */
+    colorForFeaturedResults: ComponentOptions.buildColorOption({ depend: 'showBadgeForFeaturedResults' }),
+    /**
+     * Specifies the color that should be used for "Recommended Results".
+     *
+     * This can be specified using either an hexadecimal value (eg: `#f58020`), an RGB value (eg: `rgb(125,10,36)`), or a CSS color name (eg: `pink`, `yellow`, `orange`, etc.).
+     *
+     * Default value is `undefined`, and is controlled through the default stylesheet of the framework.
+     */
+    colorForRecommendedResults: ComponentOptions.buildColorOption({ depend: 'showBadgeForRecommendedResults' })
+  };
+
+  constructor(public element: HTMLElement, public options: IPromotedResultsBadgeOptions, public bindings: IComponentBindings) {
+    super(element, PromotedResultsBadge.ID, bindings);
+    this.options = ComponentOptions.initComponentOptions(element, PromotedResultsBadge, options);
+    this.bind.onRootElement(ResultListEvents.newResultDisplayed, (args: IDisplayedNewResultEventArgs) => {
+      this.buildBadge(args.result, args.item);
```

Comment:
I see what you are trying to say, but don't you think your example is inherently more "complex" / "complicated" / "abstract" to actually understand what's going on ? 

Versus here, it seems to me it is very straightforward with no "magic" config object that abstract the actual task we want to perform on the query result/ result element.

If/when we need to add more badge and the current solution does not scale, we can look into refactoring. But for now, this seems overkill.

I can certainly extract an `appendBadge` method though, because it makes sense to extract this behaviour separately (since it is a different operation).